### PR TITLE
DMS Retry.scala retries on server errors up to a specified limit

### PR DESCRIPTION
**Pull request by nicklester**

### CODE REVIEW

File: server/cmwell-data-tools/src/test/scala/cmwell/tools/data/downloader/consumer/ConsumerSpec.scala [hunk 1]
```diff
@@ -40,6 +40,30 @@ class ConsumerSpec extends BaseWiremockSpec {
...
+      .whenScenarioStateIs(Scenario.STARTED)
+      .willReturn(aResponse()
+        .withStatus(StatusCodes.OK.intValue)
+        .withHeader(CMWELL_POSITION, "3AAAMHwv"))
+      .willSetStateTo(exitRetry)
+    )
+
+    Downloader.createTsvSource(baseUrl = s"localhost:${wireMockServer.port}")
+
+    1 shouldBe 1
```
Comment (matankdr) [inline server/cmwell-data-tools/src/test/scala/cmwell/tools/data/downloader/consumer/ConsumerSpec.scala]: You should count the number of times the wire mock server received an incoming request to `create-consumer`.
An assertion should verify that number of defined retries == actual number of incoming requests.

File: server/cmwell-data-tools/src/test/scala/cmwell/tools/data/downloader/consumer/ConsumerSpec.scala [hunk 2]
```diff
@@ -212,4 +236,23 @@ class ConsumerSpec extends BaseWiremockSpec {
...
+      .willReturn(aResponse()
+      .withStatus(StatusCodes.GatewayTimeout.intValue))
+      .willSetStateTo(Scenario.STARTED)
+    )
+
+    val result = Downloader.createTsvSource(baseUrl = s"localhost:${wireMockServer.port}").map(_ => 1)
+      .runFold(0)(_ + _)
+
+    result.flatMap(_=>{
+      1 shouldBe 1
```
Comment (matankdr) [inline server/cmwell-data-tools/src/test/scala/cmwell/tools/data/downloader/consumer/ConsumerSpec.scala]: same in here

File: server/cmwell-dc/src/pack/application.conf [hunk 3]
```diff
@@ -30,7 +30,7 @@ akka {
 data-tools {
   akka {
     http.host-connection-pool {
-      max-connections = 20
```
Comment (matankdr) [inline server/cmwell-dc/src/pack/application.conf]: are you sure you want to change this number?

File: server/cmwell-data-tools/src/test/scala/cmwell/tools/data/downloader/consumer/ConsumerSpec.scala (hunk 1 above)
Comment (nicklester) [reply server/cmwell-data-tools/src/test/scala/cmwell/tools/data/downloader/consumer/ConsumerSpec.scala]: New unit test added and working.

File: server/cmwell-dc/src/pack/application.conf (hunk 3 above)
Comment (nicklester) [reply server/cmwell-dc/src/pack/application.conf]: Agree. My bad from earlier testing. Fixed

**Comment by bryaakov** [comment]
So, where does this stand, @matankdr can you please approve?

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala [hunk 4]
```diff
@@ -113,21 +124,45 @@ object Retry extends DataToolsLogging with DataToolsConfig {
...
 
-            // schedule a retry to http stream
-            logger.warn(
-              s"$labelValue server error: will retry again in $delay host=${getHostname(h)} status=$s entity=$e"
-            )
-            val future = after(delay, system.scheduler)(Future.successful(data))
-            Some(immutable.Seq(future -> state))
+                val retryBackoff = delayWithFactor(delayFactor,c,iterationDelay,limit.get)
+
+                e.discardBytes()
```
Comment (bryaakov) [inline server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala]: Requested by @e-orz : we need to expose response body in log. We can do the trick with a "log ID" that will be printed now, and the rest once the body of the response is consumed.

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala [hunk 5]
```diff
@@ -113,21 +124,45 @@ object Retry extends DataToolsLogging with DataToolsConfig {
...
+                logger.debug(
```
Comment (bryaakov) [inline server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala]: please add response headers and response body (see above)

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala [hunk 6]
```diff
@@ -113,21 +124,45 @@ object Retry extends DataToolsLogging with DataToolsConfig {
...
+                  s"$labelValue server error - received $s, count=$count will retry again in $retryBackoff" +
+                    s" host=${getHostnameValue(h)} data=${stringifyData(data)}"
+                )
+
+                val future =
+                  after(retryBackoff, system.scheduler)(Future.successful(data))
+                Some(immutable.Seq(future -> state.copy(count = Some(c - 1), delay=retryBackoff)))
+              case Some(0) =>
+                logger.warn(
+                  s"$labelValue server error - received $s, count=$count will not not retry sending request," +
```
Comment (bryaakov) [inline server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala]: same as above - please expose headers and body

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/downloader/consumer/Downloader.scala [hunk 7]
```diff
@@ -45,6 +45,23 @@ object Downloader extends DataToolsLogging with DataToolsConfig {
...
+    FiniteDuration(timeoutDuration.length, timeoutDuration.unit)
+  }
+
+  val retryLimit = config.hasPath("cmwell.downloader.consumer.http-retry-limit") match {
+    case true => Some(config.getInt("cmwell.downloader.consumer.http-retry-limit"))
+    case false => None
+  }
+
+  val delayFactor = config.hasPath("cmwell.downloader.consumer.http-retry-delay-factor") match {
+    case true => config.getInt("cmwell.downloader.consumer.http-retry-delay-factor")
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/downloader/consumer/Downloader.scala]: Done

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala [hunk 8]
```diff
@@ -56,21 +58,30 @@ object Retry extends DataToolsLogging with DataToolsConfig {
...
                    ec: ExecutionContext,
                    label: Option[LabelId] = None) = {
 
+    implicit def asFiniteDuration(d: Duration) = scala.concurrent.duration.Duration.fromNanos(d.toNanos);
+
     val labelValue = label.map { case LabelId(id) => s"[$id]" }.getOrElse("")
     val toStrictTimeout = 30.seconds
 
+    def delayWithFactor(delayFactor: Double, countRemaining: Int, initialDelay: FiniteDuration, retryLimit : Int)  = {
+      val firstDelay = countRemaining==retryLimit
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala]: Done

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala [hunk 9]
```diff
@@ -113,21 +124,45 @@ object Retry extends DataToolsLogging with DataToolsConfig {
...
+                logger.warn(
+                  s"$labelValue server error - received $s, count=$count will not not retry sending request," +
+                    s" host=${getHostnameValue(h)} data=${stringifyData(data)}"
+                )
+                badDataLogger.info(
+                  s"$labelValue data=${concatByteStrings(data, endl).utf8String}"
+                )
+                None
+              case None =>
+                e.discardBytes()
```
[the remark this answers is not in the document]
Comment (nicklester) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala]: Done

File: server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala (hunk 4 above)
Comment (nicklester) [reply server/cmwell-data-tools/src/main/scala/cmwell/tools/data/utils/akka/Retry.scala]: Please can we discuss this some more..

**Pull request merged**